26_Autoexposure/main.cpp

```diff
@@ -0,0 +1,874 @@
+			const_cast<std::remove_const_t<decltype(m_surface)>&>(m_surface) = nbl::video::CSimpleResizeSurface<nbl::video::CDefaultSwapchainFramebuffers>::create(std::move(surface));
+		}
+		if (m_surface)
+			return { {m_surface->getSurface()/*,EQF_NONE*/} };
+		return {};
+	}
+
+	inline bool onAppInitialized(smart_refctd_ptr<ISystem>&& system) override
+	{
+		// Remember to call the base class initialization!
+		if (!device_base_t::onAppInitialized(smart_refctd_ptr(system)))
+			return false;
+		if (!asset_base_t::onAppInitialized(std::move(system)))
+			return false;
+
+		/*
+			* We'll be using a combined image sampler for this example, which lets us assign both a sampled image and a sampler to the same binding.
+			* In this example we provide a sampler at descriptor set creation time, via the SBinding struct below. This specifies that the sampler for this binding is immutable,
+			* as evidenced by the name of the field in the SBinding.
+			* Samplers for combined image samplers can also be mutable, which for a binding of a descriptor set is specified also at creation time by leaving the immutableSamplers
+			* field set to its default (nullptr).
+			*/
+		smart_refctd_ptr<IGPUDescriptorSetLayout> lumaPresentDSLayout;
+		{
+			auto defaultSampler = m_device->createSampler(
+				{
+					.AnisotropicFilter = 0
+				}
+			);
+
+			const IGPUDescriptorSetLayout::SBinding lumaPresentBindings[1] = {
+				{
+					.binding = 0,
+					.type = IDescriptor::E_TYPE::ET_COMBINED_IMAGE_SAMPLER,
+					.createFlags = IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_NONE,
+					.stageFlags = IShader::E_SHADER_STAGE::ESS_FRAGMENT | IShader::E_SHADER_STAGE::ESS_COMPUTE,
+					.count = 1,
+					.immutableSamplers = &defaultSampler
+				}
+			};
+			lumaPresentDSLayout = m_device->createDescriptorSetLayout(lumaPresentBindings);
+			if (!lumaPresentDSLayout)
+				return logFail("Failed to Create Descriptor Layout: lumaPresentDSLayout");
+		}
+
+		// Create semaphores
+		m_lumaMeterSemaphore = m_device->createSemaphore(m_submitIx);
+		m_presentSemaphore = m_device->createSemaphore(m_submitIx);
+
+		// create the descriptor sets and with enough room
+		{
+			core::smart_refctd_ptr<IDescriptorPool> lumaPresentPool;
+			{
+				const video::IGPUDescriptorSetLayout* const layouts[] = { lumaPresentDSLayout.get(), lumaPresentDSLayout.get() };
+				const uint32_t setCounts[] = { 1u, 1u };
+				lumaPresentPool = m_device->createDescriptorPoolForDSLayouts(IDescriptorPool::E_CREATE_FLAGS::ECF_NONE, layouts, setCounts);
+			}
+
+			if (!lumaPresentPool)
+				return logFail("Failed to Create Descriptor Pools");
+
+			m_lumaPresentDS[0] = lumaPresentPool->createDescriptorSet(core::smart_refctd_ptr(lumaPresentDSLayout));
+			m_lumaPresentDS[1] = lumaPresentPool->createDescriptorSet(core::smart_refctd_ptr(lumaPresentDSLayout));
+			if (!m_lumaPresentDS[0] || !m_lumaPresentDS[1])
```

Comment:
why are you double buffering descriptor sets ? the only reason old example did it was because of round robining the luma meter output SSBOs.

Now we have BDA in Push Cosntants